include/dlaf/sender/transform_mpi.h

```diff
@@ -0,0 +1,164 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <pika/mpi.hpp>
+#include <pika/unwrap.hpp>
+
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/sender/transform.h"
+#include "dlaf/sender/when_all_lift.h"
+
+namespace dlaf::comm::internal {
+
+/// Helper for moving a PromiseGuard<Communicator> into the MPI function being
+/// called by transformMPI. Callables passed to transformMPI have their
+/// arguments passed by reference, but doing so with PromiseGuard would keep the
+/// guard alive until the completion of the MPI operation, whereas we are only
+/// looking to guard the submission of the MPI operation. We therefore use this
+/// wrapper to move PromiseGuard<Communicator> into a transformMPI callables so
+/// that the guard is released on return from the callable.
+template <typename T>
+decltype(auto) movePromiseGuard(T&& t) {
+  if constexpr (std::is_same_v<std::decay_t<T>, dlaf::common::PromiseGuard<Communicator>>) {
```

Comment:
Maybe add some sort of static assert here. Everything should be coming in as a reference already...

Reply:
I changed this to only take references with `T&`.